File: src/c/testing.c
// -------------------------------------------------------------------------- //
//	testing.c
//	Performance and capacity tests of the memory algorithm
// 	https://creatingintelligence.org
// -------------------------------------------------------------------------- //
//
//	Copyright (c) 2026 Peter Overmann
//
//	SPDX-License-Identifier: MIT
//
//	This file is part of the "Creating Intelligence" C reference library.
//	It is licensed under the MIT License. You may obtain a copy of the 
//	License in the LICENSE file in the root directory of this repository.
//
//	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIN.D,
//	EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
//	MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// -------------------------------------------------------------------------- //


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <time.h>

#include "memory.h"

// -------------------------------------------------------------------------- //
//	 Test Configuration
// -------------------------------------------------------------------------- //

#define	UNKNOWN		0		// 0: Test against training data
							// 1: Test against unknown inputs

#define NOISE 		0		// Random elements added/removed (can be negative)
				
#define ITEMS		50000	// Items per iteration
#define ITERATIONS	40		// Number of iterations
#define PREFILL		0		// Items to write before testing

#define Adim		1000	// Hyperparameters
#define Apop		10
#define Bdim		1001	// Auto-associative if Adim == Bdim
#define Bpop		10


#define OPS ((int)round((double)ITEMS*CLOCKS_PER_SEC/((double)(clock()-start))))

int main(int argc, char *argv[])
	{
 	clock_t start;
 	int autoassociative = Adim == Bdim, unknowninputs = UNKNOWN;
 	
	if (autoassociative)
		printf("\n\nAuto-associative memory A->A at %d/%d, ", Adim, Apop);
	else
		printf("\n\nHetero-associative memory A->B at %d/%d and %d/%d, ", Adim, Apop, Bdim, Bpop);

	if (unknowninputs)	printf("\n   testing against unknown inputs with %d elements\n", Apop);
	else 				printf("noise = %d\n", NOISE);

	if (unknowninputs)
		printf("\n| %10s | %10s | %10s | %10s | %10s | %10s |\n|",
			"items","writes/s","memory","reads/s","false pos","overlap");
	else
		printf("\n| %10s | %10s | %10s | %10s | %10s | %10s |\n|",
			"items","writes/s","memory","reads/s","distance","overlap");
				
	for (int i = 0; i < 6; i++) printf("------------|"); printf("\n");
		
	Memory *M = Memory_new(Adim, Apop, Bdim, Bpop);
	
	Set **x = malloc(ITEMS * sizeof(Set*)),
		**y = malloc(ITEMS * sizeof(Set*)),
		**Y = malloc(ITEMS * sizeof(Set*));

	for (int i = 0; i < ITEMS; i++)
		{
		x[i] = Set_new(Adim); y[i] = Set_new(Bdim);
		Y[i] = Set_new(Bdim);
		}
						
	for (int i = 0; i < PREFILL; i++)
		{ // repurpose x[0] and y[0] instead of allocating new sets
		Set_random(x[0], Apop); Set_random(y[0], Bpop);
		if (autoassociative) Set_copy(y[0], x[0]);
		Memory_write(M, x[0], y[0]);
		}
				
	for (int iter = 1; iter <= ITERATIONS; iter ++)
		{
		// Generate test data
		for (int i = 0; i < ITEMS; i++)
			{
			Set_random(x[i], Apop); Set_random(y[i], Bpop);
			if (autoassociative) Set_copy(y[i], x[i]);
			}
		printf("| %10d |", iter * ITEMS + PREFILL); fflush(stdout);

		// Write test data
		start = clock();
		for (int i = 0; i < ITEMS; i++) Memory_write(M, x[i], y[i]);
		printf(" %10d |", OPS); fflush(stdout);
		printf(" %10.6f |",((double)Memory_count(M)) / (double)Bdim / (double)(Adim*(Adim-1)/2)); fflush(stdout);

		// Read test data
		start = clock();
		for (int i = 0; i < ITEMS; i++)
			if (unknowninputs)	Memory_read (M, Set_random(x[i], Apop), Y[i]);
			else				Memory_read (M, Set_noise(x[i], NOISE), Y[i]);

		printf(" %10d |", OPS); fflush(stdout);
		
		int H = 0, OV = 0;// Accumulated distance and overlap scores.
		for (int i = 0; i < ITEMS; i++)
			{
			if (unknowninputs) 	H += (Y[i]->p > 0); // Count non-zero responses
			else				H += distance(y[i], Y[i]);
			OV += overlap (y[i], Y[i]);
			}
		printf(" %10.6f | %10.6f |\n", ((double)H)/ITEMS, ((double)OV)/ITEMS);
		}
	Memory_free(M);
	return 0;
	}

File: experiments/testing/testing.py
"""
Copyright (c) 2026 Peter Overmann
 
SPDX-License-Identifier: MIT
 
This file is part of the "Creating Intelligence" project. It is licensed 
under the MIT License. You may obtain a copy of the License in the LICENSE 
file in the root directory of this repository.
 
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
"""

import time
import random
from creating_intelligence import Memory

def random_set(n, p):
    return sorted(random.sample(range(1, n + 1), p))

def random_association(config):
    return (
        random_set(*config["A_parameters"]),
        random_set(*config["B_parameters"])
    )

def add_noise(A, config_A, noise):
    n, p = config_A
    if noise == 0:
        return A
    elif noise > 0:
        random_ints = [random.randint(1, n) for _ in range(noise)]
        return sorted(list(set(A).union(random_ints)))
    else:
        return sorted(random.sample(A, len(A) + noise))

def overlap(A, B):
    return len(set(A).intersection(B))

def distance(A, B):
    return len(set(A).symmetric_difference(B))

def test(config):
    # Initialize memory via the configuration dictionary
    mem = Memory(config)
    
    NA, PA = mem["A_parameters"]
    NB, PB = mem["B_parameters"]
    
    memsize = (NB * NA * (NA - 1)) / 2
    size = config["batchsize"]
    
    print(f"Configuration: {config}")
    print(f"Memory: {mem}\n")
    
    cols = ["items", "writes/s", "density", "reads/s", "distance", "overlap"]
    print("|" + "".join([f"{col} ".rjust(12) + "|" for col in cols]))
    print("|" + "------------|" * len(cols))
    
    for iter_num in range(1, config["iterations"] + 1):
        data = [random_association(config) for _ in range(size)]
        
        t0 = time.perf_counter()
        for A, B in data:
            mem["store"](A, B)
        storetime = time.perf_counter() - t0
        
        inputs = [add_noise(A, config["A_parameters"], config["noise"]) for A, B in data]
        
        t1 = time.perf_counter()
        out = [mem["retrieve"](inp) for inp in inputs]
        retrievetime = time.perf_counter() - t1
        
        dist_total = sum(distance(out[i], data[i][1]) for i in range(size))
        overlap_total = sum(overlap(out[i], data[i][1]) for i in range(size))
        
        items = iter_num * size
        writes_s = round(size / storetime) if storetime > 0 else float('inf')
        density = mem["memorycount"]() / memsize
        reads_s = round(size / retrievetime) if retrievetime > 0 else float('inf')
        avg_dist = dist_total / size
        avg_overlap = overlap_total / size
        
        print(f"| {items:10d} | {writes_s:10.0f} | {density:10.6f} | {reads_s:10.0f} | {avg_dist:10.6f} | {avg_overlap:10.6f} |")

if __name__ == "__main__":
    # Unified configuration dictionary
    testconfig = {
        "A_parameters": (1000, 10),
        "B_parameters": (1001, 10),
        "batchsize": 50000,
        "iterations": 40,
        "noise": 0
    }
    
    test(testconfig)